src/bindings/jna/libelektra4j/src/test/java/org/libelektra/KeySetTest.java

```diff
@@ -158,10 +158,8 @@ public void test_keySetAppendKeySet_shouldPass() {
 	public void test_keySetCut_shouldPass() {
 		final KeySet ks = KeySet.create(6, key, key2, key3, key4, key5, key6);
 		final KeySet ks2 = ks.cut(key4);
-		assertEquals(3, ks2.length());
+		assertEquals(5, ks.length());
```

Comment:
I think you may have misunderstood, how `ksCut` works. It doesn't just remove the one key given as an argument. It actually moves all keys [below the given key]() (including the given key) into a new KeySet.

So in this test case the result should be `ks == { key, key2, key3 }` and `ks2 == { key4, key5, key6 }`.

Reply:
I asked that in the original issue #2551  and got this response: 
https://github.com/ElektraInitiative/libelektra/issues/2551#issuecomment-477610837

Reply:
The javadoc was indeed wrong, and maybe the implementation is as well, but the test case was definitely correct in its original form...

If you think of all the keys as file paths, then a KeySet is the equivalent of a file system and `ksCut` is the equivalent of cutting a directory and pasting it into a new file system (i.e. KeySet).

So the KeySet
```
/some/key
/some/key/a
/some/key/b
/some/parent/key
/some/parent/key/child
/some/parent/key/child/grandchild
/some/parent/key/otherchild
/some/parent/otherkey
/the/last/key/here
```
when cut at the key `/some/parent/key` is modified to become this KeySet
```
/some/key
/some/key/a
/some/key/b
/some/parent/otherkey
/the/last/key/here
```
and the following KeySet is created and returned
```
/some/parent/key
/some/parent/key/child
/some/parent/key/child/grandchild
/some/parent/key/otherchild
```

Here is the corresponding documentation of the C function:
https://github.com/ElektraInitiative/libelektra/blob/b641e5dc28e16e45d5dee5ffc213ab5849285682/src/libs/elektra/keyset.c#L1026-L1083

---

Also the test case should not only check that `ks2` contains the correct keys, but also that `ks` doesn't contain them anymore.

Reply:
Ok, so then I suppose that the implementation has to be fixed, because right now it's not working as described above. @markus2330 do you agree?

Reply:
May I ask why you think it is not working as described above?

I executed the test for the JNA bindings in my local copy of the repository (in the `master` branch) using:

```sh
mkdir build
cd build
cmake -GNinja -DBINDINGS="jna" ..
ctest -V -R jna
```

. The last test reports:

```
19: -------------------------------------------------------
19:  T E S T S
19: -------------------------------------------------------
19: Running org.libelektra.KeyTest
19: Tests run: 31, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.116 sec - in org.libelektra.KeyTest
19: Running org.libelektra.KeySetTest
19: Tests run: 17, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.011 sec - in org.libelektra.KeySetTest
19:
19: Results :
19:
19: Tests run: 48, Failures: 0, Errors: 0, Skipped: 0
19:
1/1 Test #19: testjna_maven ....................   Passed    2.99 sec

The following tests passed:
	testjna_maven

100% tests passed, 0 tests failed out of 1

Label Time Summary:
bindings    =   2.99 sec*proc (1 test)
kdbtests    =   2.99 sec*proc (1 test)
memleak     =   2.99 sec*proc (1 test)

Total Test time (real) =   3.04 sec
```

, which seems to indicate that the test for the JNA binding succeeds.

Now let us check, if `test_keySetCut_shouldPass` checks the behavior of `cut` correctly. Currently the tests creates a key set `ks` that contains the following keys:

```
/key_test/1/key_name
/key_test/2/key_name
/key_test/3/key_name
/key_test/4/key_name
/key_test/4/key_name/1
/key_test/4/key_name/1/123
```

. The test then “cuts” the key set on `key4` (with name `/key_test/4/key_name`) and stores the result in `ks2`. According to the documentation, after cutting the key set `ks2` should contain everything that is below `/key_test/4/key_name` (also including the cut point `/key_test/4/key_name`):

```sh
/key_test/4/key_name         # key4
/key_test/4/key_name/1       # key5
/key_test/4/key_name/1/123   # key6
```

. This is also what the test code:

```java
assertEquals(3, ks2.length());
assertNotNull(ks2.lookup(key4));
assertNotNull(ks2.lookup(key5));
assertNotNull(ks2.lookup(key6));
```

checks, as far as I can tell.

Reply:
While I was fixing issue #2574, I also noticed that there seems to be something wrong with the `assertNotNull` test statements. At least the following assertions do **not** fail:

```java
assertNotNull(ks2.lookup(key));
assertNotNull(ks2.lookup(key2));
assertNotNull(ks2.lookup(key3));
```

. I then added additional assertions to debug the situation:

```java
assertEquals(3, ks.length());
assertEquals(KEY_1_NAME, ks.at(0).getName());
assertEquals(KEY_2_NAME, ks.at(1).getName());
assertEquals(KEY_3_NAME, ks.at(2).getName());

assertEquals(3, ks2.length());
assertEquals(KEY_5_NAME, ks2.at(0).getName());
assertEquals(KEY_6_NAME, ks2.at(1).getName());
// Should this not be the first key, since 
// 
// - `/key_test/4/key_name` 
// 
// is lexically smaller than
// 
// - `/key_test/4/key_name/1`, and
// - `/key_test/4/key_name/1/123`
// 
// ?
assertEquals(KEY_4_NAME, ks2.at(2).getName());
```

. The assertions above all succeed on my machine. Apart from the ordering issue, `cut` seems to work as intended.

Reply:
Thank you for this elaborate help!

@mirunix please also add these further test cases.